ltiauthenticator/__init__.py

```diff
@@ -1,164 +1,2 @@
-
-
-        args_list = []
-        for key, values in args.items():
-            if type(values) is list:
-                args_list += [(key, value) for value in values]
-            else:
-                args_list.append((key, values))
-
-        base_string = signature.construct_base_string(
-            'POST',
-            signature.normalize_base_string_uri(launch_url),
-            signature.normalize_parameters(
-                signature.collect_parameters(body=args_list, headers=headers)
-            )
-        )
-
-        consumer_secret = self.consumers[args['oauth_consumer_key']]
-
-        sign = signature.sign_hmac_sha1(base_string, consumer_secret, None)
-        is_valid = signature.safe_string_equals(sign, args['oauth_signature'])
-
-        if not is_valid:
-            raise web.HTTPError(401, "Invalid oauth_signature")
-
-        return True
-
-
-class LTIAuthenticator(Authenticator):
-    """
-    JupyterHub Authenticator for use with LTI based services (EdX, Canvas, etc)
-    """
-
-    auto_login = True
-    login_service = 'LTI'
-
-    consumers = Dict(
-        {},
-        config=True,
-        help="""
-        A dict of consumer keys mapped to consumer secrets for those keys.
-
-        Allows multiple consumers to securely send users to this JupyterHub
-        instance.
-        """
-    )
-
-    def get_handlers(self, app):
-        return [
-            ('/lti/launch', LTIAuthenticateHandler)
-        ]
-
-
-    @gen.coroutine
-    def authenticate(self, handler, data=None):
-        # FIXME: Run a process that cleans up old nonces every other minute
-        validator = LTILaunchValidator(self.consumers)
-
-        args = {}
-        for k, values in handler.request.body_arguments.items():
-            args[k] = values[0].decode() if len(values) == 1 else [v.decode() for v in values]
-
-
-        if validator.validate_launch_request(
-                handler.request.full_url(),
-                handler.request.headers,
-                args
-        ):
-            return {
-                'name': handler.get_body_argument('user_id'),
-                'auth_state': {k: v for k, v in args.items() if not k.startswith('oauth_')}
-            }
-
-
-    def login_url(self, base_url):
-        return url_path_join(base_url, '/lti/launch')
-
-
-class LTIAuthenticateHandler(BaseHandler):
-    """
-    Handler for /lti/launch
-
-    Implements v1 of the LTI protocol for passing authentication information
-    through.
-
-    If there's a custom parameter called 'next', will redirect user to
-    that URL after authentication. Else, will send them to /home.
-    """
-
-    @gen.coroutine
-    def post(self):
-        user = yield self.login_user()
-        self.redirect(self.get_body_argument('custom_next', self.get_next_url()))
+from .auth import LTIAuthenticator
+from .validator import LTILaunchValidator, LTILaunchValidationError
```

Comment:
If you omit the `from .auth import LTIAuthenticator` here in `__init__`, instead requiring users to explicitly import `litauthenticator.validator` or `ltiauthenticator.auth`, then the validator may be used without jupyterhub being importable.

That's a backward-incompatible change in terms of imports, but a minor one.